### PR TITLE
test-backend

### PR DESCRIPTION
**Pull request by lqqyt2423**

### CODE REVIEW

**Comment by MiffyLiye** [comment]
Hello, I'm Tao from Wiredcraft, I will review this PR and ask some questions.

File: README.md [hunk 1]
```diff
@@ -19,55 +17,75 @@ Build a RESTful API that can `get/create/update/delete` user data from a persist
...
+#### Login
 
-## What We Care About
+```
+POST /login HTTP/1.1
+Content-Type: application/json
+{
+    "user": "admin",
+    "password": "123456"
+}
```
Comment (MiffyLiye) [inline README.md:58]: how to configure the user & password?

File: README.md [hunk 2]
```diff
@@ -19,55 +17,75 @@ Build a RESTful API that can `get/create/update/delete` user data from a persist
...
 
-Feel free to use any open-source library as you see fit, but remember that we are evaluating your coding skills and problem solving skills.
+Response Body Like:
+{
+    "access_token": "xxx...",
+    "token_type": "bearer",
+    "expires_in": 7200
```
Comment (MiffyLiye) [inline README.md:64]: how to configure the `expires_in` time

Comment (MiffyLiye) [reply README.md:64]: and how to configure the jwt key

File: test/user.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        authorization = `Bearer ${res.body.access_token}`
+      })
+  })
+
+  afterEach(() => {
+    sinon.restore()
+  })
+
+  it('should GET /users 200', () => {
+    sinon.stub(User, 'find').resolves([])
```
Comment (MiffyLiye) [inline test/user.js:26]: Some teams think that trying to mock database is unwise/impractical for test cases, can you compare the benefit & cost of mocking database?

File: README.md (hunk 2 above)
Comment (lqqyt2423) [reply README.md:64]: Could config in config.js, maybe get from environment variables in production.

File: test/user.js (hunk 3 above)
Comment (lqqyt2423) [reply test/user.js:26]: Mocking is simple and useful in this scenario, no need to configure a test database, no dirty data will be generated.
When testing some complex functions, use a real database would be simple and convincing.

File: controllers/user.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  const user = await User.findById(id)
+  if (!user) {
+    ctx.status = 404
+    return
+  }
+  
+  user.name = payload.name
+  user.dob = payload.dob
+  user.address = payload.address
+  user.description = payload.description
```
Comment (xavierchow) [inline controllers/user.js:69]: just for example, what if there are a massive number of fields on the user model, 
would you still prefer to go with increasing code lines with `user.xxx = payload.xxx`?

Comment (lqqyt2423) [reply controllers/user.js:69]: `Object.keys(payload).forEach(k => { user[k] = payload[k] })`
or 
`UserModel.findByIdAndUpdate(user.id, { $set: { ...payload } })`

Comment (xavierchow) [reply controllers/user.js:69]: what if some of the properties of `payload` don't need to save into the user?

Comment (lqqyt2423) [reply controllers/user.js:69]: If have undefined properties, validator will throw error.

If have other defined properties and don't need to save, can write `user.xxx = payload.xxx` line by line, it is clearer.
Or just write code like `const { otherField, ...realPayload } = payload`.

**Comment by xavierchow** [comment]
thanks, @lqqyt2423 
I'm going to close this PR and we will get back to you soon.

**Pull request closed without merging**